src/lib/logical_query_plan/compare_lqps.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "all_parameter_variant.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+class AggregateNode;
+class CreateViewNode;
+class DeleteNode;
+class DropViewNode;
+class DummyTableNode;
+class InsertNode;
+class JoinNode;
+class LimitNode;
+class PredicateNode;
+class ProjectionNode;
+class LogicalPlanRootNode;
+class ShowColumnsNode;
+class ShowTablesNode;
+class SortNode;
+class StoredTableNode;
+class UpdateNode;
+class ValidateNode;
+class MockNode;
+class UnionNode;
+class LQPColumnReference;
+class LQPExpression;
+
+class SemanticLQPCompare final {
+ public:
+  SemanticLQPCompare(const std::shared_ptr<const AbstractLQPNode>& lhs, const std::shared_ptr<const AbstractLQPNode>& rhs);
+
+  const std::vector<std::pair<std::shared_ptr<const AbstractLQPNode>, std::shared_ptr<const AbstractLQPNode>>>& differing_subtrees() const;
+
+  bool operator()();
+
+ private:
+  std::shared_ptr<const AbstractLQPNode> _lhs;
+  std::shared_ptr<const AbstractLQPNode> _rhs;
+
+  std::unordered_map<std::shared_ptr<const AbstractLQPNode>, std::shared_ptr<const AbstractLQPNode>> _node_mapping;
+
+  std::vector<std::pair<std::shared_ptr<const AbstractLQPNode>, std::shared_ptr<const AbstractLQPNode>>> _differing_subtrees;
+
+  bool _structural_traverse(const std::shared_ptr<const AbstractLQPNode>& lhs, const std::shared_ptr<const AbstractLQPNode>& rhs);
```

Comment:
Should these take shared_ptrs or references?

Reply:
Could also take references, I think. Will refactor this anyway and choose whatever fits better then.